wled00/FX.cpp

```diff
@@ -3116,6 +3116,398 @@ static uint16_t rolling_balls(void) {
+        thisHeight = 1.0f;
+        ants[i].lastBounceUpdate = strip.now;
+        ants[i].height = thisHeight;
+      }
+    }
+
+    // check if reached past the end of the strip.
+    if (thisHeight >= 1.0f && ants[i].velocity > 0.0f) {
+      if (SEGMENT.check1) {                                               // if looking for food, stop and go back the other way
+        thisHeight = 1.0f;
+        ants[i].velocity = -ants[i].velocity;                             // reverse direction
+        ants[i].lastBounceUpdate = strip.now;
+        ants[i].height = thisHeight;
+        ants[i].hasFood = false;                                          // dropped off the food, now going back for more
+        passBy = true;                                                    // when looking for food, pass by other ants without bumping into them
+        SEGMENT.check3 = true;                                            // when looking for food, pass by other ants without bumping into them
+      }
+      else {                                                              // If not looking for food, wrap around
+        thisHeight = 0.0f;
+        ants[i].lastBounceUpdate = strip.now;
+        ants[i].height = thisHeight;
+      }
+    }
+
+    // check for "passing by" or "bumping into" other ants
+    if (!passBy) {                                           // Ants bump into each other and some reverse direction if checkbox #3 (Pass by) is not "checked"; they pass each other if "checked"
+      for (int j = i + 1; j < numAnts; j++) {
+        if (ants[j].velocity != ants[i].velocity) {
+          //  tcollided + ants[j].lastBounceUpdate is acutal time of collision (this keeps precision with long to float conversions)
+          float tcollided = (cfac*(ants[i].height - ants[j].height) +
+                ants[i].velocity*float(ants[j].lastBounceUpdate - ants[i].lastBounceUpdate))/(ants[j].velocity - ants[i].velocity);
+
+          if ((tcollided > 2.0f) && (tcollided < float(strip.now - ants[j].lastBounceUpdate))) { // 2ms minimum to avoid duplicate bounces
+            ants[i].height = ants[i].height + ants[i].velocity*(tcollided + float(ants[j].lastBounceUpdate - ants[i].lastBounceUpdate))/cfac;
+            ants[j].height = ants[i].height;
+            ants[i].lastBounceUpdate = (unsigned long)(tcollided + 0.5f) + ants[j].lastBounceUpdate;
+            ants[j].lastBounceUpdate = ants[i].lastBounceUpdate;
+
+            if (ants[i].velocity > ants[j].velocity)
+              ants[i].velocity = -ants[i].velocity;
+            else
+              ants[j].velocity = -ants[j].velocity;
+            thisHeight = ants[i].height + ants[i].velocity*(strip.now - ants[i].lastBounceUpdate)/cfac;
+          }
+        }
+      }
+    }
+
+    uint32_t color;
+    if (SEGMENT.palette != 0 ) {                                                        // if a Palette is selected (besides the Default palette), use the palette's colors
```

Comment:
This `if` is unnecessary as that is what color_from_palette() does internally.
4th parameter determines which color (from the 3) it takes.